apps/prairielearn/src/question-servers/calculation-worker.js

```diff
@@ -121,14 +122,15 @@ function grade(server, coursePath, submission, variant, question) {
     v2_score: grading.score,
     feedback: grading.feedback ?? null,
     partial_scores: {},
-    submitted_answer: submission.submitted_answer ?? null,
+    submitted_answer: submission.submitted_answer ?? {},
+    raw_submitted_answer: submission.raw_submitted_answer ?? {},
     format_errors: {},
     gradable: true,
 
     // Note: v3 questions can change `params` and `true_answer` during grading, but
     // this was not implemented for v2 questions.
-    params: variant.params ?? null,
-    true_answer: variant.true_answer ?? null,
+    params: variant.params ?? {},
+    true_answer: variant.true_answer ?? {},
```

Comment:
These changes shouldn't actually have any impact on runtime. All the data here has already been through `parse(...)` `in apps/prairielearn/src/question-servers/calculation-subprocess.js`, which performs the exact same nullish coalescing. This is just needed to keep TypeScript happy. At some point, we can maybe add `NOT NULL` constraints to the underlying database columns; that should be valid.

Reply:
For my own understanding, why can't we then type variant.XXX as non-nullish?

Reply:
Consider file an issue for tracking adding that constraint too.

Reply:
Our types _must_ represent the underlying database schema exactly for our code to be safe. Currently, there's no `NOT NULL` constraint on these columns:

https://github.com/PrairieLearn/PrairieLearn/blob/99cdfbacab9052c678702d17d1a25c59ec153429/database/tables/variants.pg#L19-L21

There are a lot more columns than this that are missing non-null constraints; I don't think we'd benefit from opening issues for all of them. Generally they don't pose serious problems, we just add `NOT NULL` constraints when it's convenient.